keps/sig-node/0753-sidecarcontainers.md

```diff
@@ -0,0 +1,1561 @@
+
+Some work-arounds can be done, to alleviate the symptoms:
+ * Ignore SIGTERM on the sidecar container, so it is alive until the pod is
+   killed. This is not ideal and _greatly_ increases the time a pod will take to
+terminate. For example, if the P99 response time is 2 minutes and therefore the
+TerminationGracePeriodSeconds is set to that, the main container can finish in 2
+seconds (that might be the average) but ignoring SIGTERM in the sidecar
+container will force the pod to live for 2 minutes anyways.
+ * Use preStop hooks that just runs a "sleep X" seconds. This is very similar to
+   the previous item and has the same problems.
+
+#### Service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will be
+blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic
+functionality. These are explained in detail in the
+[alternatives](#alternatives) section. Nonetheless, here is a  quick highlight
+of some of the things some service mesh currently need to do:
+
+ * Recommend users to delay starting their apps by using a script to wait for
+   the service mesh to be ready. The goal of a service mesh to augment the app
+   functionality without modifying it, that goal is lost in this case.
+ * To guarantee that traffic goes via the services mesh, an initContainer is
+   added to blackhole traffic until the service mesh containers are up. This way,
+   other containers that might be started before the service mesh container can't
+   use the network until the service mesh container is started and ready. A side
+   effect is that traffic is blackholed until the service mesh is up and in a
+   ready state.
+ * Use preStop hooks with a "sleep infinity" to make sure the service mesh
+   doesn't terminate before other containers that might be serving requests.
+
+The auto-inject of initContainer [has caused bugs][linkerd-bug-init-last], as it
+competes with other tools auto-injecting a container to be run last too.
+
+[linkerd-bug-init-last]: https://github.com/linkerd/linkerd2/issues/4758#issuecomment-658457737
+
+### Problems: Coupling infrastructure with applications
+
+The limitations due to the lack of ordering guarantees on startup and shutdown,
+can sometimes force the coupling of application code with infrastructure. This
+causes a dependency between applications and infrastructure, and forces more
+coordination between multiple, possibly independent, teams.
+
+An example in the open source world about this is the [Istio CNI
```

Comment:
I agree it is a blurry line. Regarding service mesh, they can remove the initContainer and use CNI (at least Istio), but the sidecar proxy is always there. Maybe it is auto-injected maybe you require to be added manually, not sure what those will argue for each setup.

Do you think it is better to remove this part? I brought it up because it is an advantage for the client we are doing this for. This is an example of the open source, but they see a similar benefit with in-house service mesh that is an independent team than the ones managing the nodes.

I'm totally fine removing this part, though, if you think it is better in some way.